contracts/Bridge.sol

```diff
@@ -330,27 +340,57 @@ contract Bridge is Pausable, AccessControl {
                 });
 
             proposal._yesVotes[0] = msg.sender;
             emit ProposalCreated(chainID, _chainID, depositNonce, resourceID, dataHash);
         } else {
-            require(dataHash == proposal._dataHash, "datahash mismatch");
-            proposal._yesVotes.push(msg.sender);
-        }
+            if ((block.number).sub(proposal._proposedBlock) > _expiry) {
+                // if we number of blocks that has passed since this proposal
+                // was submitted exceeds the expiry threshold set, cancel the proposal
+                proposal._status = ProposalStatus.Cancelled;
+                emit ProposalCancelled(chainID, _chainID, depositNonce, resourceID, dataHash);
+            } else {
+                require(dataHash == proposal._dataHash, "datahash mismatch");
+                proposal._yesVotes.push(msg.sender);
 
-        _hasVotedOnProposal[chainID][depositNonce][msg.sender] = true;
-        emit ProposalVote(chainID, _chainID, depositNonce, resourceID, proposal._status);
+            }
 
-        // If _depositThreshold is set to 1, then auto finalize
-        // or if _relayerThreshold has been exceeded
-        if (_relayerThreshold <= 1 || proposal._yesVotes.length >= _relayerThreshold) {
-            proposal._status = ProposalStatus.Passed;
-            emit ProposalFinalized(chainID, _chainID, depositNonce, resourceID);
         }
+        if (proposal._status != ProposalStatus.Cancelled) {
+            _hasVotedOnProposal[chainID][depositNonce][msg.sender] = true;
+            emit ProposalVote(chainID, _chainID, depositNonce, resourceID, proposal._status);
+
+            // If _depositThreshold is set to 1, then auto finalize
+            // or if _relayerThreshold has been exceeded
+            if (_relayerThreshold <= 1 || proposal._yesVotes.length >= _relayerThreshold) {
+                proposal._status = ProposalStatus.Passed;
+                emit ProposalFinalized(chainID, _chainID, depositNonce, resourceID);
+            }
+        }
+
     }
 
+    function cancelProposal(uint8 chainID, uint64 depositNonce) external onlyRelayers {
+        Proposal storage proposal = _proposals[uint8(chainID)][depositNonce];
+        require((block.number).sub(proposal._proposedBlock) > _expiry, "Proposal does not meet expiry threshold");
+        
+        proposal._status = ProposalStatus.Cancelled;
+        emit ProposalCancelled(chainID, _chainID, depositNonce, proposal._resourceID, proposal._dataHash);
+
+    }
+
+    function adminCancelProposal(uint8 chainID, uint64 depositNonce) external onlyAdmin {
+        Proposal storage proposal = _proposals[uint8(chainID)][depositNonce];
+        require((block.number).sub(proposal._proposedBlock) > _expiry, "Proposal does not meet expiry threshold");
+
+        proposal._status = ProposalStatus.Cancelled;
+        emit ProposalCancelled(chainID, _chainID, depositNonce, proposal._resourceID, proposal._dataHash);
+    }
```

Comment:
Would it be cheaper to have an internal function that these can both call? Or perhaps combine the modifiers somehow?

Reply:
I think an internal function would be the cheaper alternative, we can't really combine the modifiers but we could maybe create a distinct `adminOrRelayer` modifier

Reply:
actually the `adminOrRelayer` modifier is `29109` gas cheaper than delegating both to an internal call, updating to that.